ambry-replication/src/main/java/com.github.ambry.replication/ReplicationManager.java

```diff
@@ -115,8 +127,89 @@ public void start() throws ReplicationException {
         this.scheduler.scheduleAtFixedRate(persistor, replicationConfig.replicationTokenFlushDelaySeconds,
             replicationConfig.replicationTokenFlushIntervalSeconds, TimeUnit.SECONDS);
       }
+      started = true;
     } catch (IOException e) {
       logger.error("IO error while starting replication", e);
+    } finally {
+      startupLatch.countDown();
+    }
+  }
+
+  /**
+   * {@inheritDoc}
+   * Note that, this method should be thread-safe because multiple threads (from different cluster change handlers) may
+   * concurrently update remote replica infos.
+   */
+  @Override
+  public void onReplicaAddedOrRemoved(List<ReplicaId> addedReplicas, List<ReplicaId> removedReplicas) {
+    // 1. wait for start() to complete
+    try {
```

Comment:
Also it looks like we have some duplicate code between this method and addReplica/removeReplica methods. Can we try to refactor?

Reply:
`addReplica()`, `removeReplica()` are methods for changing replicas on local node. They create a new PartitionInfo along with all peer RemoteReplicaInfos. `onReplicaAddedOrRemoved()` basically modifies existing PartitionInfo by adding/removing certain RemoteReplicaInfo. Refactoring will be considered in a separate PR after we have verified functionality of this change.